### PR TITLE
Added date tooltip to worldclock

### PR DESCRIPTION
**Pull request by TomaszPluta**
Hello! I would like to contribute to lxqt-panel. I added a date tooltip to worldclock plugin, as mentioned in : #1313

### CODE REVIEW

**Comment by TomaszPluta** [comment]
I will apply remarks, thanks.

**Comment by tsujan** [comment]
I know you're working on it but a simple tooltip's format doesn't need to be customizable. What's needed is only an option for showing or not showing it. The flexible format that was suggested above is enough.

**Comment by TomaszPluta** [comment]
OK, good point. I've misunderstood  "format shouldn't be hard-coded". Now it is all clear (it  shoud be taken from an method rather than be written in plain text)

**Comment by tsujan** [comment]
>  (it shoud be taken from an method rather than be written in plain text)

Exactly! In that way, it'll follow the localization automatically (different countries have different formats).

**Comment by tsujan** [comment]
Oh, before I forget it: After you clean up the PR, I'll explain in detail why and how the code should be changed. In short, we don't need to update the tooltip regularly.

**Comment by TomaszPluta** [comment]
> After you clean up the PR
Do you mean I should close this one? Should I prepare optional tooltip on my own branch/fork and then you will give me hint what could be improved (before creating pull request)?

**Comment by tsujan** [comment]
> Do you mean I should close this one?

No. Why? I see it as a work in progress. By cleanup I mean having a single option for enabling/disabling tooltip with the above-mentioned format, without an extra setting for its format. The default value of that option should disable the tooltip.

BTW, please don't change translations in a PR for the code.

**Comment by TomaszPluta** [comment]
I've added tooltip as optional. Can You review code and add next suggestions?

**Comment by tsujan** [comment]
There are 4 problems:

 1. The checkbox text is wrong; it says "ShoShow tooltip".

 2. Add `mShowTooltip(false),` below `mShowWeekNumber(true),` in LXQtWorldClock's c-tor (uninitialized variables are dangerous).

 3. In your implementation, when "Advanced manual format" is checked, there is no way to toggle "Show tooltip" because it's disabled.

 I'll explain the 4th problem and its solution after you fix the above ones because it's more important.

**Comment by TomaszPluta** [comment]
1. Fixed, letter "h" is now choosed  to combinate with "alt key for "show tooltip"'. 2. Fixed. 3. I've moved "show tooltip" option to the genreal tab, I though it better fits there (and always can be toggled).

**Comment by tsujan** [comment]
@TomaszPluta Thanks! I'll read the code tomorrow.

**Comment by tsujan** [comment]
There are 2 small issues:

  (1) The `tabstops` list isn't correct in `lxqtworldclockconfiguration.ui`. Actually, it had a problem before your patch. Please either leave it as it was before (you could fix it later, after this PR is merged) or fix it here.

  (2) The indentation of `mShowTooltip(false),` should be 4 spaces, not a tab.

OK, let me also explain the main trick here:

It isn't elegant to update a *hidden* tooltip continuously and hope that it'll be correct when shown (actually, with the current patch, it won't be correct if the clock only shows hours).

Instead, we can remove `setToolTip()` completely and install an event filter inside the c-tor:

```c++
...
...
    mMainWidget = new QWidget();
    mMainWidget->installEventFilter(this);
    mContent = new ActiveLabel();
...
...
```

Then, we just wait for `QEvent::ToolTip` in `eventFilter()` (**this needs** `#include <QToolTip>`):

```c++
bool LXQtWorldClock::eventFilter(QObject * watched, QEvent * event)
{
    if (mShowTooltip && watched == mMainWidget && event->type() == QEvent::ToolTip)
    {
        QHelpEvent *helpEvent = static_cast<QHelpEvent*>(event);
        QDateTime now = QDateTime::currentDateTime();
        QString timeZoneName = mActiveTimeZone;
        if (timeZoneName == QLatin1String("local"))
            timeZoneName = QString::fromLatin1(QTimeZone::systemTimeZoneId());
        QTimeZone timeZone(timeZoneName.toLatin1());
        QDateTime tzNow = now.toTimeZone(timeZone);
        QToolTip::showText(helpEvent->globalPos(), tzNow.toString(QLocale(QLocale::AnyLanguage, QLocale().country()).dateTimeFormat(QLocale::ShortFormat)));
        return false;
    }
    return QObject::eventFilter(watched, event);
}
```

Now — and only now — we can be sure that our tooltip is up-to-date *every time it's shown*.

File: plugin-worldclock/lxqtworldclock.h [hunk 1]
```diff
@@ -33,6 +33,7 @@
 
 #include <QDialog>
 #include <QLabel>
+#include <QToolTip>
```
Comment (tsujan) [inline plugin-worldclock/lxqtworldclock.h]: `#include <QToolTip>` can go into "lxqtworldclock.cpp"

Comment (TomaszPluta) [reply plugin-worldclock/lxqtworldclock.h]: Yes, I noticed that nothing from this include is needed in *h file

**Comment by tsujan** [comment]
@agaida This PR can be merged as it is now. It closes https://github.com/lxqt/lxqt-panel/issues/1313.

**Comment by tsujan** [comment]
Let's merge it to avoid rebasing.

**Pull request merged**